corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -873,28 +904,48 @@ def _attach_prefetch_models(objects_by_id, prefetched_models, link_field_name, c
         setattr(obj, cached_attrib_name, list(group))
 
 
-def _batch_iterate(batch_fn, next_start_from_fn, start_from=None, chunk_size=500):
+def _batch_iterate_over_shards(batch_fn, next_start_args, start_from=None, chunk_size=500):
     """
     Iterate through a function in batches. Assumes the following signatures:
 
-    batch_fn(start_from, limit) - a function that returns sorted data in batches
-    next_start_from_fn(item) - a function that returns a "start_from" for a item for the next batch
+    batch_fn(start_from, last_id, limit=limit, using=db_alias) - a function that returns sorted data in batches
+    next_start_args(item) - a function that returns a tuple of "start_from" and "last_id" for the next batch.
+    The key is used to exclude duplicate items.
     """
     start_from = start_from or datetime.min
-    # todo: this will greedily query the same data multiple times in a sharded setup. We should make it smarter
-    batch = batch_fn(start_from, limit=chunk_size)
-    while batch:
-        for item in batch:
-            yield item
-            next_start_from = next_start_from_fn(item)
-
-        if len(batch) == chunk_size:
-            # we got a full chunk so keep checking for more
-            assert next_start_from > start_from  # make sure we are making progress
-            start_from = next_start_from
-            batch = batch_fn(start_from, limit=chunk_size)
-        else:
-            batch = []  # equivalent to return
+    if not settings.USE_PARTITIONED_DATABASE:
+        db_list = [None]  # use the default database
+    else:
+        from corehq.sql_db.config import PartitionConfig
+        partition_config = PartitionConfig()
+        db_list = partition_config.get_form_processing_dbs()
+
+    db_status = {db: {'start_from': start_from, 'last_id': None, 'has_more': True} for db in db_list}
+
+    has_more = True
+    while has_more:
+        for db_alias in db_list:
```

Comment:
can't we still get things out of order even if we query one db at a time? it looks like this gets a batch from one of the shards and immediately yields what's in the batch instead of waiting for sort with what the other shards returned